doc/mgmt/OpenConfig_Path_Translation_Root_Expansion_HLD.md

```diff
@@ -0,0 +1,338 @@
+func TranslSetRequest(req *gnmipb.SetRequest) TranslatorCtx
+func TranslSetResponse(resp *gnmipb.SetResponse, ctx TranslatorCtx)
+func TranslSubscribeRequest(req *gnmipb.SubscribeRequest) TranslatorCtx
+func TranslSubscribeResponse(resp *gnmipb.SubscribeResponse, ctx TranslatorCtx)
+```
+
+#### Invocation for Translation
+##### GetRequest & GetResponse
+Translation of GetRequest/GetResponse happens inside gNMI Server’s Get process.
+
+(gnmi_server/Server.go)
+```
+// Get implements the Get RPC in gNMI specification.
+func (s *Server) Get(ctx context.Context, req *gnmipb.GetRequest) (*gnmipb.GetResponse, error) {
+  ...
+  translCtx := pathtransl.TranslGetRequest(req)
+  ... // Data clients process req, and return GetResponse resp
+  pathtransl.TranslGetResponse(resp, translCtx)
+  return resp, err
+}
+```
+
+##### SetRequest & SetResponse
+Translation of SetRequest/SetResponse happens inside GNMI Server’s Set process. 
+
+(gnmi_server/Server.go)
+```
+// Set implements the Set RPC in gNMI spec.
+func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
+  ...
+  translCtx := pathtransl.TranslSetRequest(req)
+  ... // Data clients process req, and return SetResponse resp
+  pathtransl.TranslSetResponse(resp, translCtx)
+  return resp, err
+}
+```
+
+##### SubscribeRequest
+Translation of SubscribeRequest happens inside the Subscribe Client’s Run process before request is passed to transclient.
+
+(gnmi_server/client_subscribe.go)
+```
+func (c *Client) Run(stream gnmipb.GNMI_SubscribeServer) (err error) {
+  ...
+  c.translCtx := pathtransl.TranslSubscribeRequest(req)
+  ...
+}
+```
+
+##### SubscribeResponse
+Translation of SubscribeRequest happens inside the Subscribe Client’s Send process before notification response is sent.
+
+(gnmi_server/client_subscribe.go)
+```
+func (c *Client) send(stream gnmipb.GNMI_SubscribeServer) error {
+  ...
+  pathtransl.TranslSubscribeResponse(resp, c.translCtx)
+  ...
+}
+```
+
+#### Single Root Translation
+Single Root means xpath `/openconfig`, which UMF currently does not support. We can use this translation layer to support single root as well.
+
+##### GetRequest
+Single Root GetRequest is translated into a GetRequest with all the supported OpenConfig paths as below:
+
+```
+Original GetRequest (single root)
+prefix:<origin:"openconfig">path:<>
+```
+
+```
+Translated GetRequest
+prefix:<>path:<elem:<name:"openconfig-interfaces:interfaces">>path:<elem:<name:"openconfig-lacp:lacp">>path:<elem:<name:"openconfig-platform:components">>path:<elem:<name:"openconfig-qos:qos">>path:<elem:<name:"openconfig-sampling:sampling">>path:<elem:<name:"openconfig-system:system">>
+```
+
+##### GetResponse
+Since the single root GetRequest is translated into multiple top-container paths, the JSON_IETF response will receive multiple notifications (one for each translated path). These multiple notifications need to be combined into one single JSON_IETF value and sent back to the client.
```

Comment:
Instead of translating at the gnmi_server, should we add native support in translib to handle single root requests? Aggregation seems like a costly operation, in my opinion